File: Cheater-iOS/Models/Quiz.swift
//
//  Quiz.swift
//  Cheater-iOS
//
//  Domain model for quizzes
//

import Foundation

struct Quiz: Identifiable, Codable, Sendable {
    let id: UUID
    let homeworkId: UUID
    let questions: [Question]
    let createdAt: Date

    // Classification metadata
    let topic: String?
    let subtopic: String?
    let classificationConfidence: Double?

    init(
        id: UUID = UUID(),
        homeworkId: UUID,
        questions: [Question],
        createdAt: Date = Date(),
        topic: String? = nil,
        subtopic: String? = nil,
        classificationConfidence: Double? = nil
    ) {
        self.id = id
        self.homeworkId = homeworkId
        self.questions = questions
        self.createdAt = createdAt
        self.topic = topic
        self.subtopic = subtopic
        self.classificationConfidence = classificationConfidence
    }

    // Computed properties
    var totalQuestions: Int {
        questions.count
    }

    var isValid: Bool {
        totalQuestions == 10 && questions.allSatisfy { $0.isValid }
    }

    var topicDisplay: String {
        topic?.capitalized ?? "General"
    }
}

// MARK: - Core Data Conversion
extension Quiz {
    /// Convert Core Data entity to domain model
    init?(from entity: QuizEntity) {
        guard
            let id = entity.id,
            let questionsData = entity.questionsJSON,
            let createdAt = entity.createdAt,
            let homeworkId = entity.homework?.id
        else {
            return nil
        }

        self.id = id
        self.homeworkId = homeworkId
        self.createdAt = createdAt

        // Decode questions from JSON
        do {
            let decoder = JSONDecoder()
            self.questions = try decoder.decode([Question].self, from: questionsData)
        } catch {
            print("Failed to decode questions: \(error)")
            return nil
        }
    }

    /// Update Core Data entity from domain model
    func update(entity: QuizEntity) throws {
        entity.id = self.id
        entity.createdAt = self.createdAt

        // Encode questions to JSON
        let encoder = JSONEncoder()
        entity.questionsJSON = try encoder.encode(questions)
    }

    /// Get questions JSON data for saving
    func questionsData() throws -> Data {
        let encoder = JSONEncoder()
        return try encoder.encode(questions)
    }
}

// MARK: - Claude API Response
extension Quiz {
    /// Structure for Claude API response (v2 with classification and multiple question types)
    struct ClaudeResponse: Codable, Sendable {
        let topic: String?
        let subtopic: String?
        let confidence: Double?
        let questions: [ClaudeQuestion]

        struct ClaudeQuestion: Codable, Sendable {
            let type: String?                // "mcq", "fillBlank", "shortAnswer"
            let question: String
            let options: [String]?           // Only for MCQ
            let correctIndex: Int?           // Only for MCQ
            let correctAnswer: String?       // For fillBlank and shortAnswer
            let explanation: String
        }
    }

    /// Create Quiz from Claude API response
    nonisolated init(from response: ClaudeResponse, homeworkId: UUID) {
        self.id = UUID()
        self.homeworkId = homeworkId
        self.createdAt = Date()
        self.topic = response.topic
        self.subtopic = response.subtopic
        self.classificationConfidence = response.confidence

        self.questions = response.questions.map { claudeQ in
            // Determine question type
            let questionType: QuestionType
            if let typeString = claudeQ.type {
                questionType = QuestionType(rawValue: typeString) ?? .mcq
            } else {
                // Fallback: if has options, it's MCQ
                questionType = claudeQ.options != nil ? .mcq : .fillBlank
            }

            return Question(
                type: questionType,
                question: claudeQ.question,
                options: claudeQ.options,
                correctIndex: claudeQ.correctIndex,
                correctAnswer: claudeQ.correctAnswer,
                explanation: claudeQ.explanation
            )
        }
    }
}

// MARK: - Quiz Attempt
struct QuizAttempt: Identifiable, Codable {
    let id: UUID
    let quizId: UUID
    let homeworkId: UUID
    let score: Int
    let totalQuestions: Int
    let timeTakenSeconds: Int?
    let answers: [QuestionAnswer]
    let completedAt: Date

    init(
        id: UUID = UUID(),
        quizId: UUID,
        homeworkId: UUID,
        score: Int,
        totalQuestions: Int,
        timeTakenSeconds: Int? = nil,
        answers: [QuestionAnswer],
        completedAt: Date = Date()
    ) {
        self.id = id
        self.quizId = quizId
        self.homeworkId = homeworkId
        self.score = score
        self.totalQuestions = totalQuestions
        self.timeTakenSeconds = timeTakenSeconds
        self.answers = answers
        self.completedAt = completedAt
    }

    // Computed properties
    var percentage: Int {
        guard totalQuestions > 0 else { return 0 }
        return Int((Double(score) / Double(totalQuestions)) * 100)
    }

    var passed: Bool {
        percentage >= 70
    }

    var gradeMessage: String {
        switch percentage {
        case 90...100:
            return "🎉 Excellent!"
        case 70..<90:
            return "👍 Great Job!"
        case 50..<70:
            return "👌 Good Effort!"
        default:
            return "💪 Keep Practicing!"
        }
    }
}

// MARK: - Question Answer
struct QuestionAnswer: Codable, Hashable {
    let questionId: UUID
    let questionIndex: Int
    let selectedIndex: Int?
    let correctIndex: Int
    let isCorrect: Bool
    let timeSpentSeconds: Int?

    init(
        questionId: UUID,
        questionIndex: Int,
        selectedIndex: Int?,
        correctIndex: Int,
        timeSpentSeconds: Int? = nil
    ) {
        self.questionId = questionId
        self.questionIndex = questionIndex
        self.selectedIndex = selectedIndex
        self.correctIndex = correctIndex
        self.isCorrect = selectedIndex == correctIndex
        self.timeSpentSeconds = timeSpentSeconds
    }
}

// MARK: - Sample Data
extension Quiz {
    static let sample = Quiz(
        homeworkId: UUID(),
        questions: Question.sampleList
    )
}

extension QuizAttempt {
    static let sample = QuizAttempt(
        quizId: UUID(),
        homeworkId: UUID(),
        score: 8,
        totalQuestions: 10,
        timeTakenSeconds: 154,
        answers: [
            QuestionAnswer(questionId: UUID(), questionIndex: 0, selectedIndex: 1, correctIndex: 1),
            QuestionAnswer(questionId: UUID(), questionIndex: 1, selectedIndex: 1, correctIndex: 1),
            QuestionAnswer(questionId: UUID(), questionIndex: 2, selectedIndex: 0, correctIndex: 0),
            QuestionAnswer(questionId: UUID(), questionIndex: 3, selectedIndex: 2, correctIndex: 2),
            QuestionAnswer(questionId: UUID(), questionIndex: 4, selectedIndex: 1, correctIndex: 2),
            QuestionAnswer(questionId: UUID(), questionIndex: 5, selectedIndex: 1, correctIndex: 1),
            QuestionAnswer(questionId: UUID(), questionIndex: 6, selectedIndex: 1, correctIndex: 1),
            QuestionAnswer(questionId: UUID(), questionIndex: 7, selectedIndex: 2, correctIndex: 2),
            QuestionAnswer(questionId: UUID(), questionIndex: 8, selectedIndex: 3, correctIndex: 2),
            QuestionAnswer(questionId: UUID(), questionIndex: 9, selectedIndex: 1, correctIndex: 1)
        ]
    )
}

File: Cheater-React/types/quiz.ts
/**
 * Quiz types and interfaces
 * Based on docs/01-DATA-MODELS.md
 */

import { Question, isValidQuestion } from './question';

export interface Quiz {
  id: string;                    // UUID v4
  homeworkId: string;            // Foreign key to Homework
  questions: Question[];         // Array of exactly 10 questions
  createdAt: Date;

  // Classification metadata (from Claude Vision API)
  topic?: string;                // "maths", "english", "science", "history", "generic"
  subtopic?: string;             // e.g., "algebra", "grammar", "biology", "world_war_2"
  classificationConfidence?: number; // 0.0 - 1.0
}

// Helper functions
export function getTotalQuestions(quiz: Quiz): number {
  return quiz.questions.length;
}

export function isValidQuiz(quiz: Quiz): boolean {
  return quiz.questions.length === 10 &&
         quiz.questions.every(q => isValidQuestion(q));
}

export function getTopicDisplay(quiz: Quiz): string {
  if (!quiz.topic) return 'General';

  const topicMap: Record<string, string> = {
    maths: 'Maths',
    english: 'English',
    science: 'Science',
    history: 'History',
    generic: 'General'
  };

  return topicMap[quiz.topic] || quiz.topic.charAt(0).toUpperCase() + quiz.topic.slice(1);
}

// Validation
export function validateQuiz(quiz: Partial<Quiz>): string[] {
  const errors: string[] = [];

  if (!quiz.homeworkId) {
    errors.push('Homework ID is required');
  }

  if (!quiz.questions || quiz.questions.length !== 10) {
    errors.push('Quiz must have exactly 10 questions');
  }

  if (quiz.questions) {
    quiz.questions.forEach((q, index) => {
      const questionErrors = isValidQuestion(q);
      if (!questionErrors) {
        errors.push(`Question ${index + 1}: Invalid question structure`);
      }
    });
  }

  if (quiz.classificationConfidence !== undefined) {
    if (quiz.classificationConfidence < 0 || quiz.classificationConfidence > 1) {
      errors.push('Classification confidence must be between 0 and 1');
    }
  }

  return errors;
}

// Type guard
export function isQuiz(obj: any): obj is Quiz {
  return (
    typeof obj === 'object' &&
    typeof obj.id === 'string' &&
    typeof obj.homeworkId === 'string' &&
    Array.isArray(obj.questions) &&
    obj.createdAt instanceof Date
  );
}

/**
 * Quiz Attempt - Records a completed quiz attempt
 */
export interface QuizAttempt {
  id: string;                    // UUID v4
  quizId: string;
  homeworkId: string;
  score: number;                 // Number of correct answers
  totalQuestions: number;        // Always 10
  timeTakenSeconds?: number;     // Total time in seconds
  answers: QuestionAnswer[];
  completedAt: Date;
}

export interface QuestionAnswer {
  questionId: string;
  questionIndex: number;         // 0-9
  selectedIndex?: number;        // For MCQ (0-3), undefined if skipped
  correctIndex: number;          // The correct answer index
  isCorrect: boolean;
  timeSpentSeconds?: number;
}

// Helper functions for QuizAttempt
export function calculatePercentage(attempt: QuizAttempt): number {
  if (attempt.totalQuestions === 0) return 0;
  return Math.round((attempt.score / attempt.totalQuestions) * 100);
}

export function hasPassed(attempt: QuizAttempt): boolean {
  return calculatePercentage(attempt) >= 70;
}

export function getGradeMessage(attempt: QuizAttempt): string {
  const percentage = calculatePercentage(attempt);

  if (percentage >= 90) return '🎉 Excellent!';
  if (percentage >= 70) return '👍 Great Job!';
  if (percentage >= 50) return '👌 Good Effort!';
  return '💪 Keep Practicing!';
}

export function formatTime(seconds?: number): string {
  if (!seconds) return '--:--';

  const mins = Math.floor(seconds / 60);
  const secs = seconds % 60;
  return `${mins}:${secs.toString().padStart(2, '0')}`;
}
